lib/trino-hdfs/src/main/java/io/trino/filesystem/hdfs/CachingHdfsInputFile.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.filesystem.hdfs;
+
+import alluxio.client.file.CacheContext;
+import alluxio.client.file.URIStatus;
+import alluxio.client.file.cache.CacheManager;
+import alluxio.client.file.cache.LocalCacheFileInStream;
+import alluxio.conf.AlluxioConfiguration;
+import alluxio.hadoop.AlluxioHdfsInputStream;
+import alluxio.hadoop.HdfsFileInputStream;
+import alluxio.wire.FileInfo;
+import io.trino.filesystem.Location;
+import io.trino.filesystem.TrinoInput;
+import io.trino.hdfs.CallStats;
+import io.trino.hdfs.HdfsContext;
+import io.trino.hdfs.HdfsEnvironment;
+import org.apache.hadoop.fs.FSDataInputStream;
+import org.apache.hadoop.fs.FileSystem;
+
+import java.io.IOException;
+
+import static com.google.common.hash.Hashing.md5;
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+public class CachingHdfsInputFile
+        extends HdfsInputFile
```

Comment:
Thanks for the feedback, will rework this 🙏

Reply:
Hey @raunaqmorarka 👋 I've given this a go, but I find the approach a bit uglier in practice. It leaves the HdfsInputFile untouched, but forces us to duplicate a bunch of internal variables (`environment`, `context`, `path`) and re-implement all default methods as wrapper around the HdfsFileInput.

I'm leaving it up in a separate branch at [Pluies/trino#alluxio-cache-no-inheritance](https://github.com/Pluies/trino/tree/alluxio-cache-no-inheritance) for review, and I'll defer to your final opinion on this.

Reply:
I think we definitely want to avoid inheritance and making base class variables `protected`. Composition is typically over inheritance in Trino code base.

Reply:
Fair enough, will update 👍

Reply:
This is fixed in the latest update. We only use delegation.